woocommerce/google-listings-and-ads
# Create Ads Campaign in Setup Ads Step 3

### Changes proposed in this Pull Request:

This is a follow-up / continuation from https://github.com/woocommerce/google-listings-and-ads/pull/337 and https://github.com/woocommerce/google-listings-and-ads/pull/342.

This PR moves the ads campaign creation from Step 2 to Step 3.

In Step 3, if users have not setup billing yet, they will need to click on the setup billing button and a new tab will be opened to display the Google Ads billing information setup. In the Step 3 tab, there will be billing status check API call that will be called: 

- when the tab got focus. This is done by using [window focus event](https://developer.mozilla.org/en-US/docs/Web/API/Window/focus_event).
- every 30 seconds only when the tab is in focus. This is to prevent users from seemingly getting stuck forever when they keep having the tab in focus.

When the tab is not in focus / is not active, there will be no billing status API call.

### Screenshots:

![billing status check](https://user-images.githubusercontent.com/417342/111814509-82666800-8915-11eb-82c5-b6ab6dd9954f.gif)

### Detailed test instructions:

1. Open https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsetup-ads
2. Open your browser dev tools network panel. You will use this to check billing status API call. 
2. Proceed through Step 1, Step 2 and Step 3.
3. Note that campaign is no longer created in Step 2.
3. In Step 3, if you don't have billing account setup, click on the setup billing button. It should open a new tab and show you the Google Ads billing information page.
    - When the Step 3 tab is not active / does not have focus, there will be no billing status API call.
    - When the Step 3 tab receive focus, immediately there will be a billing status API call.
    - When the Step 3 tab is in focus, the page will query the billing status every 30 seconds. 
    - When the Google Ads billing setup is completed and the billing status API call returns `status: "approved"`, it will call API to create the campaign and then redirect to the dashboard page and display the campaign creation success modal.
4. If you already have billing account setup, click on the launch campaign button, it will create the campaign, and then redirect to the dashboard page and display the campaign creation success modal.

### Important technical note with browser dev tools:

If you have your browser dev tools open, when you focus on the dev tools, the GLA app UI will lose focus, which means the 30 seconds polling will not happen.

Similarly, if you focus on the GLA app, and then focus on dev tools and focus on GLA app UI again in quick succession repeatedly, this means the GLA app has focus (call API to check billing status), lose focus, and has focus again (call API again).
